Python in production part 4 of 5 (#223)
ippontech/blog-usa: posts/python-in-production-part-4-of-5.md

---
authors:
- Lucas Ward
tags:
- python
- api
- devops
- linux
title: "Python in Production Part 4 of 5"
image: 
---

> Welcome back to the Python in Production series.  In the [third part](https://templinkplaceholder) of the series, we learned how to convert our program into a runnable binary file.  Now that we have a runnable binary, we are ready to *integrate* our program into a system.  

Reviewer: You can update the link for Part 3 now that it's published as well!

# Create a Unit File

Systemd unit files live on the file system of your linux server.  A commonly used place for *user application unit files* is `/etc/systemd/user/{file_name}.service`.  The `{file_name}` is usually the name of your program.  For our application, the full path plus the file name will be `/etc/systemd/user/simple_api.service`.  For now, we can create the file here in our local repo, and then *transfer* it to our linux server once we are finished editing it.

You should be in the top level directory of your project.  For us, that is `sample-python-program/`.  With your terminal, create the unit file.

```bash
touch simple_api.service
```

Now with your text editor, write the following code to the file.  Don't worry, we will review what each line means right after we save the file.

```ini
[Unit]
Description=Simple API Service
After=network.target

[Service]
ExecStart=/usr/local/bin/simple_api
Restart=always

[Install]
WantedBy=multi-user.target
```

> Following is an explanation of each line.

Reviewer: ```suggestion
Here is an explanation of each line.
```

* `Restart=always` tells systemd to restart our program if it closes for any reason.
* `[Install]` This section handles your service in the context of the user space.
* `WantedBy=multi-user.target` tells systemd to make our program available to multiple users on the linux system.

Now that we have our systemd unit file, we are ready to put it, along with our program binary, onto our linux server.

# Putting it on the Server

**Note:** This section involves some pretty advanced Linux topics.  It's totally cool if you don't fully understand everything.  I will offer some minor explanations, but I highly encourage you to read up on any and all commands that you run on your machines.  Please don't *blindly run* linux commands that you find on the internet.

If you have been following along since the beginning of the series, then you will have a *binary* file called `simple_api` and a *systemd unit file* called `simple_api.service` in your current working directory.  In order to get these files onto our linux server, we are going to use a utility called ***scp***.

> SCP stands for **S**ecure **C**opy.  It is a simple command line utility that transfers files from your local machine, to a remote server (or the other way around!).  If you are more comfortable with an ftp utility, or a program such as filezilla, feel free to use that instead.  

Reviewer: You've only highlighted the S & C of SCP. Isn't SCP, Secure Copy Protocol?

Reviewer: You might also want to highlight what makes the transfer secure specifically here versus just running a cp

### Copy the Binary File onto Your Linux Server

If you are using an SSH key to access your server, then run this command on your local machine, replacing the stuff in brackets with *your stuff*:

```bash
scp -i {path_to_ssh_private_key_file} simple_api {user}@{server}:/home/{user}/simple_api
```

> My command looks like this (DON'T RUN THIS ONE!): 

Reviewer: ```suggestion
My command looks like this (***DON'T RUN THIS ONE!***): 
```

Reviewer: Bolding and italicizing for extra emphasis!

```bash
scp -i ~/.ssh/id_ed25519 simple_api lward@python-blog.example.com:/home/lward/simple_api
```

If you are accessing your server via username and password, simply run this command and follow the on screen prompts for your username and password:
```bash
scp simple_api {user}@{server}:/home/{user}/simple_api
```

> An example (DON'T RUN THIS ONE!):

Reviewer: ```suggestion
An example (***DON'T RUN THIS ONE!***):
```

Reviewer: Bolding and Italicizing for extra emphasis!

```bash
scp simple_api lward@python-blog.example.com:/home/lward/simple_api
```

### Copy the SystemD Unit File onto Your Linux Server

Again, run this command from your local machine, replacing the stuff in brackets with *your stuff*:

```bash
scp -i {path_to_ssh_private_key_file} simple_api.service {user}@{server}:/home/{user}/simple_api.service
```

> The next step after copying both files on to our server, is to log into our server and complete our setup.  This next step will only need to be completed one time, and then it is good forever!

Reviewer: ```suggestion
After copying both files onto our server, the next step is to complete our setup.  This next step will only need to be completed once; it is good forever!
```

By copying the files to our user space, and then creating symlinks, it keeps the sudo operation from happening 'over the air' and instead, can be more securely ran on the actual server.  Sure you could have just moved those files from your home directory to their final destinations, but then you would have to do that *everytime* you wanted to upload a new version.  The symlinks will still be there, even if we replace our files.

## Reload SystemD and Enable your Service

Just having our files in place does not mean that systemd will just start running our application.  We have to reload the daemon to see our new unit file, and then enable and start our service.  Systemd includes `systemctl`, it is used to manage our system and user unit files.  While logged into your server, run these commands:
```bash
systemctl --user daemon-reload
systemctl --user enable simple_api.service
systemctl --user start simple_api.service
```

> These three commands tell systemd to look for new unit files, add them to the list of things to start on boot, and then start the service.  To verify our service is indeed running, we will use another piece of software included with systemd, called `journald`.  In order to read the logs stored in `journald`, we use the command `journalctl`.  You may notice a patter emerging here! 

Reviewer: patter -> pattern

# Configure Your Program With an Environment File

Let's make a slight update to our Application.  I want our application to say `{Message: Hello, World! From North Carolina}`.  I would prefer not to hard code my location, North Carolina, into our application.  In order to get around this hard coding issue, we will use an environment file. 

Switching back to our local terminal, in our project directory, let's make some changes:
```bash
touch simple_api.conf
echo "DEV_LOCATION='North Carolina'" > simple_api.conf

cd src/simple_api
```

Open up `api.py` and replace the contents with this updated version:
```python
from fastapi import FastAPI
import os

app = FastAPI()

@app.get("/")
async def read_root():
    location = os.environ.get('DEV_LOCATION', "The Underworld!")
    message = f"Hello, World! From {location}"
    return {"Message": message}
```

Next, run your build script:
```bash
cd ../..
bash build.sh
```

Let's test our application before sending it to the server.
```bash
export DEV_LOCATION="North Carolina"
./simple_api
```

> Use your browser to navigate to the site and you should see the new message `{"Message":"Hello, World! From North Carolina"}`.  When testing locally, we simply set an Environment Variable, but systemd actually has an option to load an environment file.  A few screens ago, we created a file called `simple_api.conf`.  We are going to put that file on our linux server in the `/etc/default/simple_api.conf` location.  Let's go ahead and update our unit file and then we will load it all onto our server.

Reviewer: ```suggestion
Use your browser to navigate the site and see the new message `{"Message":"Hello, World! From North Carolina"}`.  When testing locally, we set an Environment Variable, but systemd has an option to load an environment file.  Earlier, we created a file called `simple_api.conf`.  We will put that file on our Linux server in the `/etc/default/simple_api.conf` location.  Let's go ahead and update our unit file and then we will load it all onto our server.
```

scp -i {path_to_ssh_private_key_file} simple_api.service {user}@{server}:/home/{user}/simple_api.service
scp -i {path_to_ssh_private_key_file} simple_api.conf {user}@{server}:/home/{user}/simple_api.conf
```

SSH into your server and create the symlink.
```bash
ssh -A {user}@{server}
sudo ln -s /home/{user}/simple_api.conf /etc/default/simple_api.conf
```

Reload the daemon to get the update unit file, and then restart your service.
```bash
systemctl --user daemon-reload
systemctl --user restart simple_api.service
```

You should now see the updated message when using Curl.
```bash
lward@python-blog:~$ curl http://localhost:8000
{"Message":"Hello, World! From North Carolina"}
```

# Why You Should Use an Environment File vs Hard Coding

If you have made it this far, then I salute you, dedicated reader.  This series consists of 5 parts, and you are nearing the end of part 4.  Each part builds upon the previous part, and each part offers more value as far as ***lessons learned***.  To illuminate this point, I will now tell you why you are creating this configuration file versus hard coding 'North Carolina' into the source code.

> Let's say that down the road, I move to Virgina.  If I had hardcoded my location into my source code, I would need to rebuild, and relaunch my appilication to reflect this change.  Using an environment file in our systemd unit file, enables us to change aspects of our program *without* having to rebuild it.  With a tiny application like this, rebuilding would not really be that big of a struggle, but things get complicated as they get bigger.  Let's update our location to Virgina and then call it a day.

Reviewer: ```suggestion
Let's say that down the road, I move to Virginia.  If I had hardcoded my location into my source code, I would need to rebuild and relaunch my application to reflect this change.  Using an environment file in our systemd unit file enables us to change aspects of our program *without* having to rebuild it.  With a small application like this, rebuilding would be manageable, but things get complicated as applications get bigger.  Let's update our location to Virginia and then call it a day.
```
